torch/csrc/jit/api/function_impl.cpp

```diff
@@ -64,6 +65,7 @@ const c10::FunctionSchema& GraphFunction::getSchema() const {
 void preoptimizeGraph(std::shared_ptr<Graph>& graph) {
   // TODO: Invoke cleanup passes before and after inlining to reduce amount of
   // code we're copying.
+  PreInlineONNX(*graph);
```

Comment:
This shouldn't be running every time we run inlining on a graph, just as an explicitly invoked pass by ONNX before ONNX runs inlining.

Reply:
ONNX Calls the tracer which internally inlines functions. I'm trying to see where is the best place to put the pre-inline pass. Maybe somewhere in the tracer.

Reply:
When we looked at the example together, seems like tracer doesn't eagerly inline so it should be possible outside of the tracer.
```
@torch.jit.script
def fn(x):
    return x + 2
@torch.jit.script
def fn2(x):
    return fn(x) + 3
def fn3(x):
    return x + fn2(x)
traced = torch.jit.trace(fn3, (torch.rand(3, 4),))
print(traced.graph)

graph(%x : Double(3, 4)):
  %1 : Function = prim::Constant[name="fn2"]()
  %2 : Tensor = prim::CallFunction(%1, %x)
  %3 : int = prim::Constant[value=1]() # test/test_jit.py:16104:0
  %4 : Double(3, 4) = aten::add(%x, %2, %3) # test/test_jit.py:16104:0
  return (%4)


```

Reply:
Actually putting this preinline pass before calling inline does not fix the graph.
So the problem is with function->optimized_graph
This optimized_graph is actually inlined within the tracer, and it does not get updated when the function->graph is updated.
I maybe able to add an API to update the optimize_graph.

Here it is:
https://github.com/pytorch/pytorch/blob/86f3305859e7a39ab91a3ec8eb2433bc39f655b2/torch/csrc/jit/api/function_impl.h#L37

Reply:
What doesn't work if you put the pass here ? 

 https://github.com/pytorch/pytorch/blob/master/torch/onnx/utils.py#L115

Reply:
Because I'm modifying the function graph, not the function opsimized_graph. And the latter is used by the downstream code. Let me know if it's easier to have a quick call about this.